api/envoy/service/ext_proc/v3/external_processor.proto

```diff
@@ -285,9 +285,9 @@ message CommonResponse {
   // along with the CONTINUE_AND_REPLACE status.
   config.core.v3.HeaderMap trailers = 4;
 
-  // Clear the route cache for the current request.
-  // This is necessary if the remote server
-  // modified headers that are used to calculate the route.
+  // Clear the route cache for the downstream client request. This is necessary
+  // if the remote server modified headers that are used to calculate the route.
+  // This field is ignored for the upstream response message.
```

Comment:
Please add a new `RequestHeadersResponse` message and move the `clear_route_cache` flag there. It is better to express the intent through the API definition.

Reply:
Thanks for the comments!  I coded it up like below. However, this is a breaking change, and outside of the scope of this fuzzer bug fix.  I am wondering can we do it in a separate PR and give it enough time for comments.

diff --git a/api/envoy/service/ext_proc/v3/external_processor.proto b/api/envoy/service/ext_proc/v3/external_processor.proto
index 5fb9393974..9ec8964048 100644
--- a/api/envoy/service/ext_proc/v3/external_processor.proto
+++ b/api/envoy/service/ext_proc/v3/external_processor.proto
@@ -285,10 +285,19 @@ message CommonResponse {
   // along with the CONTINUE_AND_REPLACE status.
   config.core.v3.HeaderMap trailers = 4;
 
+// This message specifies the response data from the external processing server
+// specially for downstream client request. If the server set this field when
+// responding to the upstream response request, it will be ignored by Envoy.
+  DownstreamRequestResponse downstream_response = 5;
+}
+
+// This message specifies the response data from the external processing server
+// specially for downstream client request.
+message DownstreamRequestResponse {
   // Clear the route cache for the downstream client request. This is necessary
   // if the remote server modified headers that are used to calculate the route.
   // This field is ignored for the upstream response message.
-  bool clear_route_cache = 5;
+  bool clear_route_cache = 1;
 }

Reply:
filed this issue to track this API change request: https://github.com/envoyproxy/envoy/issues/27772

Reply:
Why can't the #27772 be fixed? What is the point of this intermediate change, that would need to be changed again at some point in the future?

Also it is ok to make breaking changes, since the API is still alpha.

Reply:
This breaking API change may trigger lengthy review process about documents and wording, and drag this bug fix for a little bit long.  But if you think we should do it within this PR any way, we can do that.

@stevenzzzz

Reply:
Fixing the API the right way will not be lengthy and is not larger than this PR. This speed up PR seem superfluous.

Reply:
The  proposed API change is in : https://github.com/envoyproxy/envoy/pull/27852. Even if we can push this API change through,  the logic implemented in this PR probably is still relevant, right? Like only do clear_route_cache() in decoder path.

Reply:
BTW,based on a quick look, that API change in https://github.com/envoyproxy/envoy/pull/27852 will trigger significantly code shuffling in onReceiveMessage() since encoding and decoding won't be able to share the same code, like handleHeadersResponse() or handleBodyResponse() any more.

Reply:
IMHO, we should get this PR committed to let the ext_proc fuzzer continue run and detecting new issues, if any.